In [3]:
import pandas as pd
df_fact_shipments=pd.read_csv("C:\\Users\\MY\\AppData\\Local\\Temp\\b6e77397-a60b-470c-9c06-47adcedb9ee6_global_supply_chain_dataset.zip.ee6\\fact_shipments.csv")

In [4]:
df_fact_shipments.shape

(2701, 19)

In [5]:
df_fact_shipments.head()

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
0,SHP00001,ORD100000,SUP032,SKU011,LOC05,LOC23,2025-03-26,2025-04-16,2025-05-01,Ocean,385,2626.28,10,1.0,36,85.7,Normal,Delivered,644054.95
1,SHP00002,ORD100001,SUP010,SKU055,LOC07,LOC17,2025-03-26,2025-04-10,2025-04-23,Rail,500,14755.42,5,3.0,28,22.4,Normal,Delivered,342080.00
2,SHP00003,ORD100002,SUP017,SKU011,LOC07,LOC20,2025-11-26,2025-12-09,2025-12-12,Rail,290,3073.81,1,3.0,16,24.8,Normal,Delivered,485132.30
3,SHP00004,ORD100003,SUP002,SKU052,LOC14,LOC25,2026-02-10,2026-03-01,2026-03-03,Ocean,371,17989.84,4,0.0,21,46.2,Normal,Delivered,94653.23
4,SHP00005,ORD100004,SUP014,SKU024,LOC05,LOC16,2025-01-23,2025-02-14,2025-02-12,Ocean,348,4490.58,0,0.0,20,28.8,Normal,Delivered,103530.00


In [6]:
df_fact_shipments.info()

<class 'pandas.DataFrame'>
RangeIndex: 2701 entries, 0 to 2700
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Shipment_ID           2701 non-null   str    
 1   Order_ID              2701 non-null   str    
 2   Supplier_ID           2700 non-null   str    
 3   Product_ID            2701 non-null   str    
 4   Origin_ID             2701 non-null   str    
 5   Destination_ID        2700 non-null   str    
 6   Ship_Date             2701 non-null   str    
 7   Expected_Date         2700 non-null   str    
 8   Actual_Delivery_Date  2701 non-null   str    
 9   Shipping_Mode         2701 non-null   str    
 10  Quantity              2701 non-null   int64  
 11  Freight_Cost_USD      2700 non-null   float64
 12  Port_Dwell_Days       2701 non-null   int64  
 13  Customs_Delay_Days    2700 non-null   float64
 14  Transit_Time_Days     2701 non-null   int64  
 15  Supplier_Risk_Score   2700 non-n

In [7]:
df_fact_shipments.columns

Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD'],
      dtype='str')

In [8]:
df_fact_shipments.isnull().sum()

Shipment_ID             0
Order_ID                0
Supplier_ID             1
Product_ID              0
Origin_ID               0
Destination_ID          1
Ship_Date               0
Expected_Date           1
Actual_Delivery_Date    0
Shipping_Mode           0
Quantity                0
Freight_Cost_USD        1
Port_Dwell_Days         0
Customs_Delay_Days      1
Transit_Time_Days       0
Supplier_Risk_Score     1
Priority_Level          0
Shipment_Status         0
Revenue_Exposure_USD    0
dtype: int64

In [9]:
df_fact_shipments.duplicated().sum()

np.int64(0)

In [10]:
df_fact_shipments["Shipping_Mode"].unique()

<StringArray>
['Ocean', 'Rail', 'Air', 'ocean', 'OCEAN']
Length: 5, dtype: str

In [11]:
df_fact_shipments.describe()

,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Revenue_Exposure_USD
count,2701.000000,2700.000000,2701.000000,2700.000000,2701.000000,2700.000000,2701.000000
mean,256.454276,11160.385637,4.428360,2.483333,24.596816,46.118593,243181.016475
std,141.001281,15326.329965,2.512147,2.215698,8.863030,18.202194,198251.384094
min,-15.000000,32.160000,0.000000,0.000000,0.000000,-5.500000,1041.900000
25%,132.000000,2577.780000,3.000000,1.000000,18.000000,33.375000,81067.190000
50%,260.000000,6237.405000,4.000000,2.000000,25.000000,46.400000,183776.400000
75%,378.000000,13379.570000,6.000000,4.000000,32.000000,58.700000,367999.520000
max,500.000000,359187.120000,14.000000,65.000000,48.000000,116.400000,858184.820000


In [12]:
df_fact_shipments["Shipping_Mode"].value_counts()

Shipping_Mode
Ocean    1599
Air       673
Rail      427
ocean       1
OCEAN       1
Name: count, dtype: int64

In [13]:
df_fact_shipments[df_fact_shipments["Shipping_Mode"].isin(["ocean","OCEAN"])]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
250,SHP00251,ORD100250,SUP017,SKU016,LOC06,LOC24,2026-05-20,2026-06-13,2026-06-15,ocean,33,1189.92,1,1.0,26,17.9,Priority,Delivered,48069.45
251,SHP00252,ORD100251,SUP038,SKU022,LOC08,LOC20,2025-08-05,2025-08-10,2025-08-17,OCEAN,498,87465.51,6,1.0,12,65.7,Normal,Delivered,736083.84


In [14]:
df_fact_shipments["Shipping_Mode"]=df_fact_shipments["Shipping_Mode"].str.strip().str.title()

In [15]:
df_fact_shipments["Shipping_Mode"].value_counts()

Shipping_Mode
Ocean    1601
Air       673
Rail      427
Name: count, dtype: int64

In [16]:
shipment_id_count=df_fact_shipments["Shipment_ID"].value_counts()
shipment_id_count[shipment_id_count > 1]

Shipment_ID
SHP00101    2
Name: count, dtype: int64

In [17]:
df_fact_shipments[df_fact_shipments["Shipment_ID"]=="SHP00101"]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
100,SHP00101,ORD100100,SUP009,SKU036,LOC04,LOC21,2025-11-06,2025-11-26,2025-11-25,Ocean,415,17977.23,0,2.0,19,20.7,Priority,Delivered,355742.15
101,SHP00101,ORD100101,SUP013,SKU011,LOC13,LOC19,2025-04-18,2025-04-29,2025-05-04,Rail,128,1374.27,2,2.0,16,46.7,Normal,Delivered,214127.36


In [18]:
df_fact_shipments[df_fact_shipments["Shipment_ID"].isin(["SHP001001","SHP00101","SHP00102","SHP00103","SHP00104"])][["Shipment_ID","Order_ID"]]

,Shipment_ID,Order_ID
100,SHP00101,ORD100100
101,SHP00101,ORD100101
102,SHP00103,ORD100102
103,SHP00104,ORD100103


In [19]:
df_fact_shipments.loc[df_fact_shipments["Order_ID"]=="ORD100101","Shipment_ID"]="SHP00102"

In [20]:
df_fact_shipments[df_fact_shipments["Shipment_ID"].isin(["SHP001001","SHP00101","SHP00102","SHP00103","SHP00104"])][["Shipment_ID","Order_ID"]]

,Shipment_ID,Order_ID
100,SHP00101,ORD100100
101,SHP00102,ORD100101
102,SHP00103,ORD100102
103,SHP00104,ORD100103


In [21]:
shipment_counts=df_fact_shipments["Shipment_ID"].value_counts()
shipment_counts[shipment_counts>1]

Series([], Name: count, dtype: int64)

#### Data-quality fixes

Standardized Shipping_Mode: ocean / OCEAN → Ocean

Found a duplicate Shipment_ID: SHP00101

Determined the two rows were actually different shipments

Corrected the second shipment's ID to SHP00102

Verified that no duplicate Shipment IDs remain

In [22]:
df_fact_shipments.isnull().sum()

Shipment_ID             0
Order_ID                0
Supplier_ID             1
Product_ID              0
Origin_ID               0
Destination_ID          1
Ship_Date               0
Expected_Date           1
Actual_Delivery_Date    0
Shipping_Mode           0
Quantity                0
Freight_Cost_USD        1
Port_Dwell_Days         0
Customs_Delay_Days      1
Transit_Time_Days       0
Supplier_Risk_Score     1
Priority_Level          0
Shipment_Status         0
Revenue_Exposure_USD    0
dtype: int64

In [23]:
df_fact_shipments[df_fact_shipments["Supplier_ID"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
330,SHP00331,ORD100330,NaN,SKU006,LOC13,LOC15,2025-05-13,2025-06-12,2025-06-11,Ocean,135,1802.25,2,0.0,29,62.1,Normal,Delivered,190262.25


In [24]:
df_fact_shipments[df_fact_shipments["Product_ID"]=="SKU006"]["Supplier_ID"].value_counts()

Supplier_ID
SUP030    5
SUP024    4
SUP017    4
SUP011    3
SUP040    3
SUP033    2
SUP034    2
SUP006    2
SUP031    1
SUP026    1
SUP010    1
SUP008    1
SUP021    1
SUP019    1
SUP003    1
SUP028    1
SUP022    1
SUP038    1
SUP020    1
SUP004    1
SUP023    1
SUP018    1
SUP032    1
SUP035    1
SUP025    1
SUP037    1
SUP005    1
Name: count, dtype: int64

In [25]:
df_fact_shipments[df_fact_shipments["Supplier_ID"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
330,SHP00331,ORD100330,NaN,SKU006,LOC13,LOC15,2025-05-13,2025-06-12,2025-06-11,Ocean,135,1802.25,2,0.0,29,62.1,Normal,Delivered,190262.25


In [26]:
df_fact_shipments[(df_fact_shipments["Product_ID"]=="SKU006")&(df_fact_shipments["Origin_ID"]=="LOC13")&(df_fact_shipments["Destination_ID"]=="LOC15")&(df_fact_shipments["Shipping_Mode"]=="Ocean")][["Supplier_ID","Product_ID","Origin_ID","Destination_ID","Shipping_Mode"]]

,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Shipping_Mode
330,NaN,SKU006,LOC13,LOC15,Ocean


In [27]:
df_fact_shipments[(df_fact_shipments["Product_ID"]=="SKU006")&(df_fact_shipments["Origin_ID"]=="LOC13")&(df_fact_shipments["Destination_ID"]=="LOC15")][["Supplier_ID","Product_ID","Origin_ID","Destination_ID"]]

,Supplier_ID,Product_ID,Origin_ID,Destination_ID
330,NaN,SKU006,LOC13,LOC15


In [28]:
df_fact_shipments[(df_fact_shipments["Product_ID"]=="SKU006")&(df_fact_shipments["Origin_ID"]=="LOC13")&(df_fact_shipments["Shipping_Mode"]=="Ocean")][["Supplier_ID","Product_ID","Origin_ID","Shipping_Mode"]]

,Supplier_ID,Product_ID,Origin_ID,Shipping_Mode
330,NaN,SKU006,LOC13,Ocean
455,SUP019,SKU006,LOC13,Ocean
1779,SUP004,SKU006,LOC13,Ocean


In [29]:
df_fact_shipments[(df_fact_shipments["Product_ID"]=="SKU006")&(df_fact_shipments["Destination_ID"]=="LOC15")&(df_fact_shipments["Shipping_Mode"]=="Ocean")][["Supplier_ID","Product_ID","Destination_ID","Shipping_Mode"]]

,Supplier_ID,Product_ID,Destination_ID,Shipping_Mode
330,NaN,SKU006,LOC15,Ocean
426,SUP008,SKU006,LOC15,Ocean
748,SUP003,SKU006,LOC15,Ocean
2112,SUP035,SKU006,LOC15,Ocean


In [30]:
df_fact_shipments[(df_fact_shipments["Origin_ID"]=="LOC13")&(df_fact_shipments["Destination_ID"]=="LOC15")&(df_fact_shipments["Shipping_Mode"]=="Ocean")][["Supplier_ID","Origin_ID","Destination_ID","Shipping_Mode"]]

,Supplier_ID,Origin_ID,Destination_ID,Shipping_Mode
72,SUP017,LOC13,LOC15,Ocean
297,SUP018,LOC13,LOC15,Ocean
330,NaN,LOC13,LOC15,Ocean
963,SUP017,LOC13,LOC15,Ocean
2040,SUP031,LOC13,LOC15,Ocean
2237,SUP019,LOC13,LOC15,Ocean
2365,SUP027,LOC13,LOC15,Ocean
2568,SUP017,LOC13,LOC15,Ocean
2637,SUP011,LOC13,LOC15,Ocean


In [31]:
df_fact_shipments.isnull().sum()

Shipment_ID             0
Order_ID                0
Supplier_ID             1
Product_ID              0
Origin_ID               0
Destination_ID          1
Ship_Date               0
Expected_Date           1
Actual_Delivery_Date    0
Shipping_Mode           0
Quantity                0
Freight_Cost_USD        1
Port_Dwell_Days         0
Customs_Delay_Days      1
Transit_Time_Days       0
Supplier_Risk_Score     1
Priority_Level          0
Shipment_Status         0
Revenue_Exposure_USD    0
dtype: int64

In [32]:
df_fact_shipments[df_fact_shipments["Destination_ID"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
417,SHP00418,ORD100417,SUP037,SKU002,LOC11,NaN,2026-03-05,2026-03-25,2026-03-27,Ocean,484,13704.11,3,1.0,22,36.8,Critical,Delivered,339172.68


In [33]:
df_fact_shipments[(df_fact_shipments["Order_ID"]=="ORD100417")]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
417,SHP00418,ORD100417,SUP037,SKU002,LOC11,NaN,2026-03-05,2026-03-25,2026-03-27,Ocean,484,13704.11,3,1.0,22,36.8,Critical,Delivered,339172.68


In [34]:
df_fact_shipments[(df_fact_shipments["Supplier_ID"]=="SUP037")&(df_fact_shipments["Product_ID"]=="SKU002")]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
417,SHP00418,ORD100417,SUP037,SKU002,LOC11,NaN,2026-03-05,2026-03-25,2026-03-27,Ocean,484,13704.11,3,1.0,22,36.8,Critical,Delivered,339172.68
724,SHP00725,ORD100724,SUP037,SKU002,LOC06,LOC17,2026-09-23,2026-10-03,2026-10-09,Air,141,19835.76,3,5.0,16,38.2,Normal,Delivered,98808.57


In [35]:
df_fact_shipments[df_fact_shipments["Expected_Date"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
1730,SHP01731,ORD101730,SUP019,SKU013,LOC01,LOC21,2026-10-27,NaN,2026-11-29,Ocean,133,1894.8,4,5.0,33,51.8,Critical,Delivered,221849.32


In [36]:
df_fact_shipments[(df_fact_shipments["Ship_Date"]=="2026-10-27")&(df_fact_shipments["Destination_ID"]=="LOC21")]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
1730,SHP01731,ORD101730,SUP019,SKU013,LOC01,LOC21,2026-10-27,NaN,2026-11-29,Ocean,133,1894.8,4,5.0,33,51.8,Critical,Delivered,221849.32


In [37]:
df_fact_shipments[(df_fact_shipments["Supplier_ID"]=="SUP019")&(df_fact_shipments["Product_ID"]=="SKU013")]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
1730,SHP01731,ORD101730,SUP019,SKU013,LOC01,LOC21,2026-10-27,NaN,2026-11-29,Ocean,133,1894.80,4,5.0,33,51.8,Critical,Delivered,221849.32
2008,SHP02009,ORD102008,SUP019,SKU013,LOC03,LOC24,2026-10-09,2026-10-18,2026-10-20,Air,61,2970.02,7,0.0,11,68.8,Normal,Delivered,101750.44
2237,SHP02238,ORD102237,SUP019,SKU013,LOC13,LOC15,2026-05-30,2026-06-29,2026-07-05,Ocean,164,2576.73,8,3.0,36,67.9,Normal,Delivered,273558.56


In [38]:
df_fact_shipments[df_fact_shipments["Freight_Cost_USD"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
588,SHP00589,ORD100588,SUP004,SKU034,LOC11,LOC21,2026-08-31,2026-09-29,2026-10-09,Ocean,393,NaN,9,5.0,39,37.4,Priority,Delivered,555301.14


In [39]:
df_fact_shipments[(df_fact_shipments["Product_ID"]=="SKU034")&(df_fact_shipments["Quantity"]==393)][[
    "Supplier_ID",
    "Product_ID",
    "Shipping_Mode",
    "Quantity",
    "Freight_Cost_USD"
]]

,Supplier_ID,Product_ID,Shipping_Mode,Quantity,Freight_Cost_USD
588,SUP004,SKU034,Ocean,393,NaN


In [40]:
df_fact_shipments[(df_fact_shipments["Supplier_ID"]=="SUP004")&(df_fact_shipments["Shipping_Mode"]=="Ocean")&(df_fact_shipments["Priority_Level"]=="Priority")][[
    "Supplier_ID",
    "Product_ID",
    "Shipping_Mode",
    "Quantity",
    "Freight_Cost_USD"
]]

,Supplier_ID,Product_ID,Shipping_Mode,Quantity,Freight_Cost_USD
60,SUP004,SKU056,Ocean,410,3287.92
137,SUP004,SKU019,Ocean,313,13449.53
189,SUP004,SKU005,Ocean,161,4014.28
588,SUP004,SKU034,Ocean,393,NaN
1054,SUP004,SKU051,Ocean,95,3829.13
1070,SUP004,SKU047,Ocean,280,6503.24
1188,SUP004,SKU046,Ocean,323,10838.51
1214,SUP004,SKU030,Ocean,264,2796.34
1280,SUP004,SKU013,Ocean,343,3793.19
1443,SUP004,SKU034,Ocean,313,10512.16


In [41]:
df_fact_shipments[df_fact_shipments["Customs_Delay_Days"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
1290,SHP01291,ORD101290,SUP008,SKU014,LOC01,LOC20,2025-10-05,2025-10-26,2025-10-30,Rail,452,23727.75,6,NaN,25,22.3,Normal,Delivered,523248.76


In [42]:
df_fact_shipments[df_fact_shipments["Supplier_Risk_Score"].isnull()]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
1510,SHP01511,ORD101510,SUP040,SKU023,LOC02,LOC17,2025-06-12,2025-07-11,2025-07-17,Ocean,437,10951.81,5,4.0,35,NaN,Priority,Delivered,131768.61


In [43]:
df_dim_supplier=pd.read_csv(r"C:\Users\MY\AppData\Local\Temp\15a16588-9818-4e68-a9ff-b3745dd9ad50_global_supply_chain_dataset.zip.d50\dim_supplier.csv")

In [44]:
df_dim_supplier[df_dim_supplier["Supplier_ID"]=="SUP040"]

,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
39,SUP040,BluePeak Electronics,Accessories,Mexico,North America,53,26,0.878


In [45]:
df_fact_shipments.loc[df_fact_shipments["Order_ID"]=="ORD101510","Supplier_Risk_Score"]=53

In [46]:
df_fact_shipments[df_fact_shipments["Order_ID"]=="ORD101510"][["Order_ID","Supplier_Risk_Score"]]

,Order_ID,Supplier_Risk_Score
1510,ORD101510,53.0


#### Column	treatment till now
| Column                | Missing | What we decided                                                             |
| --------------------- | ------: | --------------------------------------------------------------------------- |
| `Supplier_ID`         |       1 | Could not reliably recover the supplier → **left missing**                  |
| `Destination_ID`      |       1 | Could not determine the destination with enough evidence → **left missing** |
| `Expected_Date`       |       1 | Could not reliably reconstruct the expected date → **left missing**         |
| `Freight_Cost_USD`    |       1 | No consistent basis to estimate the cost → **left missing**                 |
| `Customs_Delay_Days`  |       1 | Couldn't assume missing = 0 → **left missing**                              |
| `Supplier_Risk_Score` |   1 → 0 | Recovered `53` from `dim_supplier` using `Supplier_ID = SUP040` → **fixed** |


In [47]:
df_fact_shipments.dtypes

Shipment_ID                 str
Order_ID                    str
Supplier_ID                 str
Product_ID                  str
Origin_ID                   str
Destination_ID              str
Ship_Date                   str
Expected_Date               str
Actual_Delivery_Date        str
Shipping_Mode               str
Quantity                  int64
Freight_Cost_USD        float64
Port_Dwell_Days           int64
Customs_Delay_Days      float64
Transit_Time_Days         int64
Supplier_Risk_Score     float64
Priority_Level              str
Shipment_Status             str
Revenue_Exposure_USD    float64
dtype: object

In [48]:
date_cols=["Ship_Date","Expected_Date","Actual_Delivery_Date"]

for col in date_cols:
    df_fact_shipments[col]=pd.to_datetime(df_fact_shipments[col],errors="coerce")

In [49]:
df_fact_shipments[date_cols].dtypes

Ship_Date               datetime64[us]
Expected_Date           datetime64[us]
Actual_Delivery_Date    datetime64[us]
dtype: object

In [50]:
df_fact_shipments.isnull().sum()

Shipment_ID             0
Order_ID                0
Supplier_ID             1
Product_ID              0
Origin_ID               0
Destination_ID          1
Ship_Date               0
Expected_Date           1
Actual_Delivery_Date    0
Shipping_Mode           0
Quantity                0
Freight_Cost_USD        1
Port_Dwell_Days         0
Customs_Delay_Days      1
Transit_Time_Days       0
Supplier_Risk_Score     0
Priority_Level          0
Shipment_Status         0
Revenue_Exposure_USD    0
dtype: int64

#### Datatype correction

The three date columns were initially stored as strings (object):

Ship_Date
Expected_Date
Actual_Delivery_Date

converted them to proper datetime values

In [51]:
df_fact_shipments.columns

Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD'],
      dtype='str')

In [52]:
df_fact_shipments[df_fact_shipments["Actual_Delivery_Date"]<df_fact_shipments["Ship_Date"]][["Shipment_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Port_Dwell_Days","Customs_Delay_Days","Transit_Time_Days","Shipment_Status"]]


,Shipment_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Shipment_Status
710,SHP00711,2025-01-21,2025-02-14,2025-01-19,0,0.0,25,Delivered


In [53]:
df_fact_shipments["Calculated_time_days"]=(df_fact_shipments["Actual_Delivery_Date"]- df_fact_shipments["Ship_Date"]).dt.days

df_fact_shipments[["Calculated_time_days","Transit_Time_Days"]].head(20)

,Calculated_time_days,Transit_Time_Days
0,36,36
1,28,28
2,16,16
3,21,21
4,20,20
5,36,36
6,42,42
7,35,35
8,28,28
9,17,17


In [54]:
(df_fact_shipments["Calculated_time_days"]==df_fact_shipments["Transit_Time_Days"]).value_counts()

True     2699
False       2
Name: count, dtype: int64

In [55]:
df_fact_shipments[df_fact_shipments["Calculated_time_days"] != df_fact_shipments["Transit_Time_Days"]][["Shipment_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Transit_Time_Days"]]

,Shipment_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Transit_Time_Days
710,SHP00711,2025-01-21,2025-02-14,2025-01-19,25
835,SHP00836,2025-12-17,2026-01-05,2026-01-10,0


In [56]:
df_fact_shipments["Calculated_time_days"]=(df_fact_shipments["Actual_Delivery_Date"]- df_fact_shipments["Ship_Date"]).dt.days

df_fact_shipments[["Calculated_time_days","Transit_Time_Days"]][df_fact_shipments["Shipment_ID"]=="SHP00836"]

,Calculated_time_days,Transit_Time_Days
835,24,0


In [57]:
df_fact_shipments.loc[df_fact_shipments["Shipment_ID"]=="SHP00836","Transit_Time_Days"]=24

In [58]:
df_fact_shipments["Calculated_time_days"]=(df_fact_shipments["Actual_Delivery_Date"]- df_fact_shipments["Ship_Date"]).dt.days

df_fact_shipments[["Calculated_time_days","Transit_Time_Days"]][df_fact_shipments["Shipment_ID"]=="SHP00836"]

,Calculated_time_days,Transit_Time_Days
835,24,24


In [59]:
df_fact_shipments.loc[df_fact_shipments["Shipment_ID"]=="SHP00711","Actual_Delivery_Date"]="2025-02-15"

In [60]:
df_fact_shipments[df_fact_shipments["Shipment_ID"]=="SHP00711"][["Shipment_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Transit_Time_Days"]]

,Shipment_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Transit_Time_Days
710,SHP00711,2025-01-21,2025-02-14,2025-02-15,25


In [61]:
df_fact_shipments.columns

Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD',
       'Calculated_time_days'],
      dtype='str')

In [62]:
df_fact_shipments.drop(columns=["Calculated_time_days"], inplace=True)

In [63]:
calculated=(df_fact_shipments["Actual_Delivery_Date"]- df_fact_shipments["Ship_Date"]).dt.days

(calculated==df_fact_shipments["Transit_Time_Days"]).value_counts()

True    2701
Name: count, dtype: int64

#### Date/business-rule validation

SHP00711

Actual delivery was before shipment.
Transit_Time_Days = 25
Corrected Actual_Delivery_Date to 2025-02-15.

SHP00836

Dates indicated 24 days.
Transit_Time_Days incorrectly contained 0.
Corrected it to 24.

In [64]:
df_fact_shipments[df_fact_shipments["Quantity"]<=0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
920,SHP00921,ORD100920,SUP016,SKU039,LOC02,LOC16,2026-04-16,2026-05-13,2026-05-17,Ocean,-15,6516.9,2,4.0,31,39.9,Normal,Delivered,322064.88


In [65]:
df_fact_shipments[(df_fact_shipments["Supplier_ID"]=="SUP016")&(df_fact_shipments["Product_ID"]=="SKU039")][[
    "Shipment_ID","Supplier_ID","Product_ID","Shipping_Mode","Quantity","Revenue_Exposure_USD"
]]

,Shipment_ID,Supplier_ID,Product_ID,Shipping_Mode,Quantity,Revenue_Exposure_USD
920,SHP00921,SUP016,SKU039,Ocean,-15,322064.88
1850,SHP01851,SUP016,SKU039,Rail,158,112331.68


In [66]:
df_fact_shipments[df_fact_shipments["Quantity"]==0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [67]:
df_fact_shipments[df_fact_shipments["Freight_Cost_USD"]<=0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [68]:
df_fact_shipments[df_fact_shipments["Port_Dwell_Days"]<0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [69]:
df_fact_shipments[df_fact_shipments["Customs_Delay_Days"]<0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [70]:
df_fact_shipments[df_fact_shipments["Transit_Time_Days"]<0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [71]:
df_fact_shipments[df_fact_shipments["Supplier_Risk_Score"]<0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD
294,SHP00295,ORD100294,SUP022,SKU015,LOC04,LOC18,2026-05-06,2026-05-15,2026-05-18,Air,406,39909.52,4,3.0,12,-0.5,Normal,Delivered,706740.44
646,SHP00647,ORD100646,SUP022,SKU029,LOC05,LOC25,2025-07-04,2025-07-30,2025-07-26,Ocean,159,505.60,0,0.0,22,-1.2,Normal,Delivered,57801.27
968,SHP00969,ORD100968,SUP035,SKU036,LOC12,LOC16,2026-10-05,2026-10-14,2026-10-17,Air,498,98986.60,5,1.0,12,-0.3,Normal,Delivered,426890.58
1236,SHP01237,ORD101236,SUP017,SKU053,LOC07,LOC19,2026-05-12,2026-06-01,2026-06-11,Ocean,412,10571.86,7,4.0,30,-2.6,Normal,Delivered,321219.92
2006,SHP02007,ORD102006,SUP035,SKU010,LOC11,LOC19,2025-01-04,2025-01-19,2025-01-20,Ocean,77,2439.94,1,0.0,16,-5.5,Normal,Delivered,100141.58


In [72]:
df_dim_supplier[df_dim_supplier["Supplier_ID"].isin(["SUP022","SUP035","SUP017"])][["Supplier_ID","Supplier_Name","Supplier_Risk_Score"]]

,Supplier_ID,Supplier_Name,Supplier_Risk_Score
16,SUP017,Zenith Electronics,27
21,SUP022,Delta Micro Systems,16
34,SUP035,Skyline Tech Supply,16


In [73]:
df_fact_shipments.loc[df_fact_shipments["Supplier_ID"]=="SUP017","Supplier_Risk_Score"]=27
df_fact_shipments.loc[df_fact_shipments["Supplier_ID"]=="SUP022","Supplier_Risk_Score"]=16
df_fact_shipments.loc[df_fact_shipments["Supplier_ID"]=="SUP035","Supplier_Risk_Score"]=16



In [74]:
df_fact_shipments[df_fact_shipments["Supplier_Risk_Score"]<0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


In [75]:
df_fact_shipments[df_fact_shipments["Revenue_Exposure_USD"]<=0]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD


#### Numeric & Business-Rule Validation Summary

Freight Cost:	No negative/zero values; 1 null	Null left unresolved

Operational Delays:	No problematic zero/negative values found	No changes

Supplier Risk Score:	5 negative values found	Corrected using dim_supplier

Revenue Exposure:	No zero/negative values	No changes

Unresolved issue: SHP00921 has Quantity = -15; no reliable replacement was found.

In [76]:
df_fact_shipments[df_fact_shipments["Actual_Delivery_Date"]>df_fact_shipments["Expected_Date"]][[
    "Shipment_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Transit_Time_Days","Shipment_Status"
    ]]

,Shipment_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Transit_Time_Days,Shipment_Status
0,SHP00001,2025-03-26,2025-04-16,2025-05-01,36,Delivered
1,SHP00002,2025-03-26,2025-04-10,2025-04-23,28,Delivered
2,SHP00003,2025-11-26,2025-12-09,2025-12-12,16,Delivered
3,SHP00004,2026-02-10,2026-03-01,2026-03-03,21,Delivered
5,SHP00006,2026-04-19,2026-05-17,2026-05-25,36,Delivered
...,...,...,...,...,...,...
2695,SHP02696,2026-01-02,2026-01-31,2026-02-08,37,Delivered
2696,SHP02697,2025-10-10,2025-11-05,2025-11-14,35,Delivered
2697,SHP02698,2026-06-22,2026-07-04,2026-07-05,13,Delivered
2698,SHP02699,2026-07-06,2026-07-25,2026-08-05,30,Delivered


In [77]:
df_fact_shipments[df_fact_shipments["Actual_Delivery_Date"]<=df_fact_shipments["Expected_Date"]][[
    "Shipment_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Transit_Time_Days","Shipment_Status"
    ]]

,Shipment_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Transit_Time_Days,Shipment_Status
4,SHP00005,2025-01-23,2025-02-14,2025-02-12,20,Delivered
25,SHP00026,2025-03-18,2025-04-05,2025-04-05,18,Delivered
35,SHP00036,2025-03-15,2025-04-07,2025-04-05,21,Delivered
37,SHP00038,2026-11-21,2026-12-04,2026-12-04,13,Delivered
43,SHP00044,2025-03-04,2025-03-23,2025-03-23,19,Delivered
...,...,...,...,...,...,...
2659,SHP02660,2025-01-17,2025-02-14,2025-02-12,26,Delivered
2666,SHP02667,2026-02-22,2026-03-24,2026-03-24,30,Delivered
2679,SHP02680,2026-10-15,2026-11-01,2026-11-01,17,Delivered
2686,SHP02687,2025-12-29,2026-01-30,2026-01-30,32,Delivered


In [78]:
import numpy as np
df_fact_shipments["Delivery_Performance"]=np.select(
    [
        df_fact_shipments["Expected_Date"].isna(),
        df_fact_shipments["Actual_Delivery_Date"]>df_fact_shipments["Expected_Date"]
    ],
    [
        "unknown",
        "late"
    ],default="on_time"
)

In [79]:
df_fact_shipments.columns

Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD',
       'Delivery_Performance'],
      dtype='str')

In [80]:
df_fact_shipments["Delivery_Performance"].value_counts(dropna=False)

Delivery_Performance
late       2460
on_time     240
unknown       1
Name: count, dtype: int64

#### Delivery Performance — Summary

Compared Actual_Delivery_Date with Expected_Date.

If Actual > Expected → Late

If Actual ≤ Expected → On Time

One shipment had a missing Expected_Date → Unknown

Created a new column: Delivery_Performance

Used np.select() and .isna() to handle the three categories.


In [81]:
df_fact_shipments[
    ~df_fact_shipments["Supplier_ID"].isin(
    df_dim_supplier["Supplier_ID"]
)][["Shipment_ID","Supplier_ID"]]

,Shipment_ID,Supplier_ID
330,SHP00331,NaN


In [82]:
df_fact_shipments[
    df_fact_shipments["Supplier_ID"].notna()&
    ~ df_fact_shipments["Supplier_ID"].isin(
        df_dim_supplier["Supplier_ID"]
    )
    ]

,Shipment_ID,Order_ID,Supplier_ID,Product_ID,Origin_ID,Destination_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Shipping_Mode,Quantity,Freight_Cost_USD,Port_Dwell_Days,Customs_Delay_Days,Transit_Time_Days,Supplier_Risk_Score,Priority_Level,Shipment_Status,Revenue_Exposure_USD,Delivery_Performance


### Table by Table

In [83]:
df_dim_date=pd.read_csv(r"C:\Users\MY\AppData\Local\Temp\a7f8c841-4190-4252-b312-8c8b9531e97c_global_supply_chain_dataset.zip.97c\dim_date.csv")

df_dim_location=pd.read_csv(r"C:\Users\MY\AppData\Local\Temp\9e3ac4bc-ed66-43ae-bb51-e71c678497b5_global_supply_chain_dataset.zip.7b5\dim_location.csv")

df_dim_product=pd.read_csv(r"C:\Users\MY\AppData\Local\Temp\38853a6d-155f-43fc-a7f0-48c28331ebeb_global_supply_chain_dataset.zip.beb\dim_product.csv")

df_fact_inventory=pd.read_csv(r"C:\Users\MY\AppData\Local\Temp\d5921c30-5415-4ba9-8457-80014fdb67b7_global_supply_chain_dataset.zip.7b7\fact_inventory.csv")

In [84]:
df_dim_supplier.shape

(40, 8)

In [85]:
df_dim_supplier.info()

<class 'pandas.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Supplier_ID          40 non-null     str    
 1   Supplier_Name        39 non-null     str    
 2   Supplier_Category    40 non-null     str    
 3   Country              40 non-null     str    
 4   Region               40 non-null     str    
 5   Supplier_Risk_Score  40 non-null     int64  
 6   Avg_Lead_Time_Days   40 non-null     int64  
 7   On_Time_Rate         40 non-null     float64
dtypes: float64(1), int64(2), str(5)
memory usage: 2.6 KB


In [86]:
df_dim_supplier.head()

,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
0,SUP001,GlobalTech Components,Components,Taiwan,APAC,29,20,0.946
1,SUP002,Nordic Parts Ltd,Accessories,South Korea,APAC,52,16,0.876
2,SUP003,Pacific Devices,Packaging,Singapore,Southeast Asia,67,23,0.848
3,SUP004,EuroCore GmbH,Packaging,Bangladesh,South Asia,32,18,0.892
4,SUP005,Shenzhen Advanced Electronics,electronics,Thailand,Southeast Asia,49,17,0.865


In [87]:
df_dim_supplier.isna().sum()

Supplier_ID            0
Supplier_Name          1
Supplier_Category      0
Country                0
Region                 0
Supplier_Risk_Score    0
Avg_Lead_Time_Days     0
On_Time_Rate           0
dtype: int64

In [88]:
df_dim_supplier[df_dim_supplier["Supplier_Name"].isna()]

,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
35,SUP036,NaN,Mechanical,Sri Lanka,South Asia,69,17,0.812


In [89]:
df_dim_supplier[df_dim_supplier["Supplier_ID"]=="SUP036"][["Supplier_ID","Supplier_Name"]]

,Supplier_ID,Supplier_Name
35,SUP036,NaN


In [90]:
df_dim_supplier[df_dim_supplier["Supplier_Category"]=="Mechanical"][["Supplier_ID","Supplier_Name","Supplier_Category"]]

,Supplier_ID,Supplier_Name,Supplier_Category
10,SUP011,BlueWave Semiconductors,Mechanical
14,SUP015,MetroTech Manufacturing,Mechanical
17,SUP018,Silverline Components,Mechanical
19,SUP020,Nova Industrial,Mechanical
21,SUP022,Delta Micro Systems,Mechanical
22,SUP023,Crestline Electronics,Mechanical
23,SUP024,Titan Components,Mechanical
24,SUP025,Evergreen Devices,Mechanical
28,SUP029,Quantum Parts,Mechanical
34,SUP035,Skyline Tech Supply,Mechanical


In [91]:
df_dim_supplier.loc[df_dim_supplier["Supplier_ID"]=="SUP036","Supplier_Name"]="Unknown"
df_dim_supplier.loc[df_dim_supplier["Supplier_ID"]=="SUP036"]

,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
35,SUP036,Unknown,Mechanical,Sri Lanka,South Asia,69,17,0.812


In [92]:
df_dim_supplier["Supplier_Category"].value_counts()

Supplier_Category
Mechanical       11
Electronics       9
Accessories       7
Components        6
Packaging         5
electronics       1
 Electronics      1
Name: count, dtype: int64

In [93]:
df_dim_supplier["Supplier_Category"]=df_dim_supplier["Supplier_Category"].str.strip().str.title()
df_dim_supplier["Supplier_Category"].value_counts()

Supplier_Category
Electronics    11
Mechanical     11
Accessories     7
Components      6
Packaging       5
Name: count, dtype: int64

In [94]:
df_dim_supplier["Country"].value_counts()


Country
Mexico          5
Singapore       4
USA             4
Taiwan          3
Thailand        2
Japan           2
Italy           2
Sweden          2
Saudi Arabia    2
Netherlands     2
Germany         2
Sri Lanka       2
South Korea     1
Bangladesh      1
germany         1
Australia       1
Canada          1
Vietnam         1
China           1
Poland          1
Name: count, dtype: int64

In [95]:
df_dim_supplier["Region"].value_counts()


Region
North America     9
Europe            9
APAC              8
Southeast Asia    8
South Asia        3
Middle East       2
EUROPE            1
Name: count, dtype: int64

In [96]:
df_dim_supplier.columns

Index(['Supplier_ID', 'Supplier_Name', 'Supplier_Category', 'Country',
       'Region', 'Supplier_Risk_Score', 'Avg_Lead_Time_Days', 'On_Time_Rate'],
      dtype='str')

In [97]:
df_dim_supplier[df_dim_supplier["Supplier_Risk_Score"]<0]

,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate


In [98]:
df_dim_supplier[df_dim_supplier["Avg_Lead_Time_Days"]<0]


,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
20,SUP021,Horizon Tech,Components,Vietnam,Southeast Asia,58,-5,0.862


In [99]:
df_fact_shipments[(df_fact_shipments["Supplier_ID"]=="SUP021")&(df_fact_shipments["Supplier_Risk_Score"]==58)][["Shipment_ID","Supplier_ID","Ship_Date","Expected_Date","Actual_Delivery_Date","Transit_Time_Days"]]

,Shipment_ID,Supplier_ID,Ship_Date,Expected_Date,Actual_Delivery_Date,Transit_Time_Days
912,SHP00913,SUP021,2025-04-21,2025-05-01,2025-05-04,13


In [100]:
df_dim_supplier[(df_dim_supplier["On_Time_Rate"]<0) | (df_dim_supplier["On_Time_Rate"]>1)]


,Supplier_ID,Supplier_Name,Supplier_Category,Country,Region,Supplier_Risk_Score,Avg_Lead_Time_Days,On_Time_Rate
31,SUP032,CoreLink Devices,Packaging,Japan,APAC,95,23,1.4


In [101]:
df_fact_shipments.columns


Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD',
       'Delivery_Performance'],
      dtype='str')

In [102]:
df_fact_shipments[df_fact_shipments["Supplier_ID"]=="SUP032"][["Shipment_ID","Supplier_ID","Expected_Date","Actual_Delivery_Date","Customs_Delay_Days","Transit_Time_Days","Delivery_Performance"]]

,Shipment_ID,Supplier_ID,Expected_Date,Actual_Delivery_Date,Customs_Delay_Days,Transit_Time_Days,Delivery_Performance
0,SHP00001,SUP032,2025-04-16,2025-05-01,1.0,36,late
81,SHP00082,SUP032,2026-06-27,2026-07-05,5.0,18,late
82,SHP00083,SUP032,2026-08-25,2026-09-05,9.0,29,late
88,SHP00089,SUP032,2025-12-03,2025-12-09,1.0,28,late
269,SHP00270,SUP032,2025-09-21,2025-09-30,5.0,21,late
383,SHP00384,SUP032,2025-04-11,2025-04-12,4.0,29,late
432,SHP00433,SUP032,2025-02-25,2025-03-06,3.0,13,late
484,SHP00485,SUP032,2025-03-04,2025-03-09,1.0,30,late
589,SHP00590,SUP032,2025-09-09,2025-09-25,3.0,31,late
601,SHP00602,SUP032,2026-10-09,2026-10-20,0.0,39,late


In [103]:
df_dim_supplier.isna().sum()

Supplier_ID            0
Supplier_Name          0
Supplier_Category      0
Country                0
Region                 0
Supplier_Risk_Score    0
Avg_Lead_Time_Days     0
On_Time_Rate           0
dtype: int64

In [104]:
df_dim_supplier["Supplier_ID"].duplicated().sum()

np.int64(0)

####  Checked for data quality issues
Covered:

Missing values 

Category formatting 

Negative risk scores 

Negative lead time : identified and documented

Invalid on-time rate : identified and documented

Duplicate supplier IDs : 0 

In [105]:
df_dim_product.shape

(60, 7)

In [106]:
df_dim_product.info()

<class 'pandas.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Product_ID              60 non-null     str    
 1   SKU                     60 non-null     str    
 2   Product_Name            60 non-null     str    
 3   Category                60 non-null     str    
 4   Unit_Cost_USD           60 non-null     float64
 5   Unit_Selling_Price_USD  59 non-null     float64
 6   Criticality             60 non-null     str    
dtypes: float64(2), str(5)
memory usage: 3.4 KB


In [107]:
df_dim_product.head()

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
0,SKU001,LAP-001,Business Laptop,Computing,634.66,599.46,Medium
1,SKU002,MON-002,4K Monitor,Displays,802.45,700.77,High
2,SKU003,CMP-003,CPU Module,Components,333.39,1689.11,High
3,SKU004,NET-004,Network Adapter,computing,169.24,1503.60,High
4,SKU005,ACC-005,Webcam,Accessories,729.27,1579.86,High


In [108]:
df_dim_product.duplicated().sum()

np.int64(0)

In [109]:
df_dim_product["Product_ID"].duplicated().sum()


np.int64(1)

In [110]:
df_dim_product[df_dim_product["Product_ID"].duplicated(keep=False)]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
9,SKU010,CMP-010,Memory Module,Components,969.85,1300.54,Medium
44,SKU010,CMP-045,Motherboard,Components,136.74,1318.63,High


In [111]:
df_dim_product.iloc[43:46]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
43,SKU044,MON-044,Ultrawide Monitor,Displays,841.39,1555.56,High
44,SKU010,CMP-045,Motherboard,Components,136.74,1318.63,High
45,SKU046,NET-046,Wireless Mouse,Networking,1124.73,286.15,Low


In [112]:
df_dim_product[df_dim_product["Product_ID"].isin(["SKU010","SKU045"])]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
9,SKU010,CMP-010,Memory Module,Components,969.85,1300.54,Medium
44,SKU010,CMP-045,Motherboard,Components,136.74,1318.63,High


In [113]:
df_dim_product[df_dim_product["SKU"]=="CMP-045"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
44,SKU010,CMP-045,Motherboard,Components,136.74,1318.63,High


In [114]:
df_fact_shipments[
    df_fact_shipments["Product_ID"] == "SKU045"
][[
    "Shipment_ID",
    "Order_ID",
    "Product_ID",
    "Quantity",
    "Revenue_Exposure_USD"
]]

,Shipment_ID,Order_ID,Product_ID,Quantity,Revenue_Exposure_USD
111,SHP00112,ORD100111,SKU045,382,503716.66
172,SHP00173,ORD100172,SKU045,439,578878.57
224,SHP00225,ORD100224,SKU045,433,570966.79
249,SHP00250,ORD100249,SKU045,255,336250.65
451,SHP00452,ORD100451,SKU045,109,143730.67
500,SHP00501,ORD100500,SKU045,201,265044.63
607,SHP00608,ORD100607,SKU045,393,518221.59
686,SHP00687,ORD100686,SKU045,375,494486.25
789,SHP00790,ORD100789,SKU045,226,298010.38
793,SHP00794,ORD100793,SKU045,318,419324.34


In [115]:
df_dim_product[df_dim_product["Product_ID"] == "SKU045"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality


In [116]:
df_dim_product.loc[df_dim_product["SKU"]=="CMP-045","Product_ID"]="SKU045"
df_dim_product["Product_ID"].duplicated().sum()

np.int64(0)

##### Duplicate Product_ID identified and corrected. SKU010 was assigned to both Memory Module and Motherboard. The Motherboard record contained SKU = CMP-045, while SKU045 was referenced by shipment records, supporting correction of its Product_ID to SKU045.

In [117]:
df_dim_product["SKU"].duplicated().sum()

np.int64(1)

In [118]:
df_dim_product[df_dim_product["SKU"].duplicated(keep=False)]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
8,SKU009,MON-009,1TB SSD,Displays,840.76,1064.86,High
46,SKU047,MON-009,Mechanical Keyboard,Accessories,722.58,1418.51,Medium


In [119]:
df_dim_product.iloc[44:48]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
44,SKU045,CMP-045,Motherboard,Components,136.74,1318.63,High
45,SKU046,NET-046,Wireless Mouse,Networking,1124.73,286.15,Low
46,SKU047,MON-009,Mechanical Keyboard,Accessories,722.58,1418.51,Medium
47,SKU048,SSD-048,Enterprise SSD,Storage,1136.58,489.69,Medium


In [120]:
df_dim_product[df_dim_product["Category"]=="Accessories"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
4,SKU005,ACC-005,Webcam,Accessories,729.27,1579.86,High
11,SKU012,ACC-012,Mechanical Keyboard,Accessories,378.61,902.46,Medium
18,SKU019,ACC-019,Webcam,Accessories,1044.59,1103.45,Medium
25,SKU026,ACC-026,Mechanical Keyboard,Accessories,49.85,142.22,High
32,SKU033,ACC-033,Webcam,Accessories,1016.20,1629.16,Low
39,SKU040,ACC-040,Wireless Mouse,Accessories,373.83,1710.25,Low
46,SKU047,MON-009,Mechanical Keyboard,Accessories,722.58,1418.51,Medium
53,SKU054,ACC-054,Mechanical Keyboard,Accessories,813.70,630.53,Low


In [121]:
df_dim_product[df_dim_product["SKU"]=="ACC-047"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality


In [122]:
df_dim_product.loc[df_dim_product["Product_ID"]=="SKU047","SKU"]="ACC-047"
df_dim_product[df_dim_product["Product_ID"]=="SKU047"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
46,SKU047,ACC-047,Mechanical Keyboard,Accessories,722.58,1418.51,Medium


In [123]:
df_dim_product["SKU"].duplicated().sum()

np.int64(0)

#### Current dim_product findings
Duplicate Product_ID → fixed (SKU010 / SKU045)

Duplicate SKU → fixed (MON-009 → ACC-047)

In [124]:
df_dim_product.isnull().sum()

Product_ID                0
SKU                       0
Product_Name              0
Category                  0
Unit_Cost_USD             0
Unit_Selling_Price_USD    1
Criticality               0
dtype: int64

In [125]:
df_dim_product["Product_Name"].value_counts()

Product_Name
CPU Module             4
Network Adapter        4
Mechanical Keyboard    4
Workstation            4
Enterprise SSD         4
4K Monitor             3
Webcam                 3
Mini PC                3
Surge Protector        3
Ultrawide Monitor      3
Business Laptop        2
External SSD           2
UPS 2000VA             2
Firewall Appliance     2
1TB SSD                2
UPS 1000VA             2
Wireless Mouse         2
 1TB SSD               1
Memory Module          1
NAS Drive              1
24-inch Monitor        1
Portable HDD           1
2TB SSD                1
WiFi Router            1
Power Adapter          1
Motherboard            1
27-inch Monitor        1
Enterprise Switch      1
Name: count, dtype: int64

In [126]:
df_dim_product["Product_Name"]=df_dim_product["Product_Name"].str.strip().str.title()

In [127]:
df_dim_product["Product_Name"].value_counts()

Product_Name
Cpu Module             4
Network Adapter        4
Mechanical Keyboard    4
Workstation            4
Enterprise Ssd         4
4K Monitor             3
Webcam                 3
Mini Pc                3
1Tb Ssd                3
Surge Protector        3
Ultrawide Monitor      3
Business Laptop        2
External Ssd           2
Ups 2000Va             2
Firewall Appliance     2
Ups 1000Va             2
Wireless Mouse         2
Memory Module          1
Nas Drive              1
24-Inch Monitor        1
Portable Hdd           1
2Tb Ssd                1
Wifi Router            1
Power Adapter          1
Motherboard            1
27-Inch Monitor        1
Enterprise Switch      1
Name: count, dtype: int64

In [128]:
df_dim_product["Category"].value_counts()

Category
Computing      9
Displays       9
Components     9
Accessories    8
Storage        8
Power          8
Networking     8
computing      1
Name: count, dtype: int64

In [129]:
df_dim_product["Category"]=df_dim_product["Category"].str.strip().str.title()
df_dim_product["Category"].value_counts()

Category
Computing      10
Displays        9
Components      9
Accessories     8
Storage         8
Power           8
Networking      8
Name: count, dtype: int64

In [130]:
df_dim_product[df_dim_product["Unit_Cost_USD"]<0]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
15,SKU016,MON-016,24-Inch Monitor,Displays,-50.0,1456.65,Medium


In [131]:
df_dim_product[df_dim_product["Product_Name"]=="24-Inch Monitor"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
15,SKU016,MON-016,24-Inch Monitor,Displays,-50.0,1456.65,Medium


In [132]:
df_dim_product[df_dim_product["Category"]=="Displays"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
1,SKU002,MON-002,4K Monitor,Displays,802.45,700.77,High
8,SKU009,MON-009,1Tb Ssd,Displays,840.76,1064.86,High
15,SKU016,MON-016,24-Inch Monitor,Displays,-50.00,1456.65,Medium
22,SKU023,MON-023,4K Monitor,Displays,723.46,NaN,Low
29,SKU030,MON-030,4K Monitor,Displays,318.03,1047.51,Medium
36,SKU037,MON-037,Ultrawide Monitor,Displays,338.09,290.29,Medium
43,SKU044,MON-044,Ultrawide Monitor,Displays,841.39,1555.56,High
50,SKU051,MON-051,Ultrawide Monitor,Displays,964.11,908.73,Medium
57,SKU058,MON-058,27-Inch Monitor,Displays,1124.96,1414.43,Medium


In [133]:
df_dim_product[df_dim_product["Unit_Selling_Price_USD"].isna()]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
22,SKU023,MON-023,4K Monitor,Displays,723.46,NaN,Low


In [134]:
df_dim_product[df_dim_product["Category"]=="Displays"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
1,SKU002,MON-002,4K Monitor,Displays,802.45,700.77,High
8,SKU009,MON-009,1Tb Ssd,Displays,840.76,1064.86,High
15,SKU016,MON-016,24-Inch Monitor,Displays,-50.00,1456.65,Medium
22,SKU023,MON-023,4K Monitor,Displays,723.46,NaN,Low
29,SKU030,MON-030,4K Monitor,Displays,318.03,1047.51,Medium
36,SKU037,MON-037,Ultrawide Monitor,Displays,338.09,290.29,Medium
43,SKU044,MON-044,Ultrawide Monitor,Displays,841.39,1555.56,High
50,SKU051,MON-051,Ultrawide Monitor,Displays,964.11,908.73,Medium
57,SKU058,MON-058,27-Inch Monitor,Displays,1124.96,1414.43,Medium


In [135]:
df_dim_product[(df_dim_product["Unit_Selling_Price_USD"]<df_dim_product["Unit_Cost_USD"])]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
0,SKU001,LAP-001,Business Laptop,Computing,634.66,599.46,Medium
1,SKU002,MON-002,4K Monitor,Displays,802.45,700.77,High
6,SKU007,PWR-007,Ups 2000Va,Power,194.08,69.46,High
17,SKU018,NET-018,Firewall Appliance,Networking,961.26,453.28,Medium
20,SKU021,PWR-021,Surge Protector,Power,558.33,479.96,High
23,SKU024,CMP-024,1Tb Ssd,Components,389.64,297.50,Low
24,SKU025,NET-025,Network Adapter,Networking,359.55,314.59,High
31,SKU032,NET-032,Wifi Router,Networking,1004.31,136.19,HIGH
35,SKU036,LAP-036,Workstation,Computing,1032.65,857.21,Medium
36,SKU037,MON-037,Ultrawide Monitor,Displays,338.09,290.29,Medium


In [136]:
df_dim_product["Criticality"].value_counts()

Criticality
Medium    27
High      20
Low       12
HIGH       1
Name: count, dtype: int64

In [137]:
df_dim_product["Criticality"]=df_dim_product["Criticality"].str.strip().str.title()
df_dim_product["Criticality"].value_counts()

Criticality
Medium    27
High      21
Low       12
Name: count, dtype: int64

#### dim_product – Cleaning Summary

Checked missing values → none

Fixed duplicate Product_ID → SKU010 / SKU045

Fixed duplicate SKU → MON-009 / ACC-047

Found negative cost for SKU016 → left unresolved

Checked selling price < cost → treated as business condition

Standardized Criticality → High/Medium/Low

Standardized Category → consistent formatting

Verified Product_ID duplicates → 0

Verified SKU duplicates → 0

In [138]:
df_dim_location.shape

(25, 5)

In [139]:
df_dim_location.info()

<class 'pandas.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Location_ID    25 non-null     str  
 1   Location_Name  25 non-null     str  
 2   Location_Type  25 non-null     str  
 3   Country        25 non-null     str  
 4   Region         25 non-null     str  
dtypes: str(5)
memory usage: 1.1 KB


In [140]:
df_dim_location.head()

,Location_ID,Location_Name,Location_Type,Country,Region
0,LOC01,Shanghai Port,Port,China,APAC
1,LOC02,Shenzhen Port,Port,China,APAC
2,LOC03,Singapore Port,Port,Singapore,Southeast Asia
3,LOC04,Rotterdam Port,Port,Netherlands,Europe
4,LOC05,Hamburg Port,Port,Germany,Europe


In [141]:
df_dim_location.isna().sum()

Location_ID      0
Location_Name    0
Location_Type    0
Country          0
Region           0
dtype: int64

In [142]:
df_dim_location["Location_Type"].duplicated().sum()

np.int64(22)

In [143]:
df_dim_location["Location_Type"].value_counts()

Location_Type
Port         14
Warehouse     7
Hub           4
Name: count, dtype: int64

In [144]:
df_dim_location["Country"].value_counts()

Country
UAE            4
India          4
USA            4
Singapore      3
Germany        3
China          2
Netherlands    2
South Korea    1
Japan          1
Sri Lanka      1
Name: count, dtype: int64

In [145]:
df_dim_location["Region"].value_counts()

Region
Europe            5
South Asia        5
APAC              4
Middle East       4
North America     4
Southeast Asia    3
Name: count, dtype: int64

#### dim_location status

Missing values → None 

Duplicate Location_ID → None 

Country consistency → No anomalies 

Region consistency → No anomalies

In [146]:
df_dim_date.shape

(730, 9)

In [147]:
df_dim_date.info()

<class 'pandas.DataFrame'>
RangeIndex: 730 entries, 0 to 729
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Date        730 non-null    str  
 1   Year        730 non-null    int64
 2   Month       730 non-null    int64
 3   Month_Name  730 non-null    str  
 4   Quarter     730 non-null    str  
 5   Week        730 non-null    int64
 6   Day         730 non-null    int64
 7   Day_Name    730 non-null    str  
 8   Is_Weekend  730 non-null    bool 
dtypes: bool(1), int64(4), str(4)
memory usage: 46.5 KB


In [148]:
df_dim_date.isna().sum()

Date          0
Year          0
Month         0
Month_Name    0
Quarter       0
Week          0
Day           0
Day_Name      0
Is_Weekend    0
dtype: int64

In [149]:
df_dim_date.head()

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend
0,2025-01-01,2025,1,January,Q1,1,1,Wednesday,False
1,2025-01-02,2025,1,January,Q1,1,2,Thursday,False
2,2025-01-03,2025,1,January,Q1,1,3,Friday,False
3,2025-01-04,2025,1,January,Q1,1,4,Saturday,True
4,2025-01-05,2025,1,January,Q1,1,5,Sunday,True


In [150]:
df_dim_date.duplicated().sum()

np.int64(0)

In [151]:
df_dim_date["Date"].min(),df_dim_date["Date"].max()

('2025-01-01', '2026-12-31')

In [152]:
df_dim_date["Date"]=pd.to_datetime(df_dim_date["Date"],errors="coerce")
df_dim_date["Date"].dtypes,df_dim_date["Date"].isna().sum()

(dtype('<M8[us]'), np.int64(0))

In [153]:
df_dim_date["Date"].min(),df_dim_date["Date"].max()

(Timestamp('2025-01-01 00:00:00'), Timestamp('2026-12-31 00:00:00'))

In [154]:
df_dim_date.columns

Index(['Date', 'Year', 'Month', 'Month_Name', 'Quarter', 'Week', 'Day',
       'Day_Name', 'Is_Weekend'],
      dtype='str')

In [155]:
df_dim_date[df_dim_date["Year"]!=df_dim_date["Date"].dt.year]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [156]:
df_dim_date[df_dim_date["Month"]!=df_dim_date["Date"].dt.month]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [157]:
df_dim_date[df_dim_date["Month_Name"]!=df_dim_date["Date"].dt.month_name()]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [158]:
df_dim_date[df_dim_date["Quarter"] != "Q" + df_dim_date["Date"].dt.quarter.astype(str)]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [159]:
df_dim_date[df_dim_date["Week"] != df_dim_date["Date"].dt.isocalendar().week]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [160]:
df_dim_date[df_dim_date["Day"] != df_dim_date["Date"].dt.day]

,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [161]:
df_dim_date[df_dim_date["Day_Name"] != df_dim_date["Date"].dt.day_name()]


,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


In [162]:
df_dim_date[df_dim_date["Is_Weekend"] != df_dim_date["Date"].dt.dayofweek.isin([5,6])]


,Date,Year,Month,Month_Name,Quarter,Week,Day,Day_Name,Is_Weekend


#### Validation completed:

Year 

Month 

Month_Name 

Quarter ← check

Week ← check

Day 

Day_Name 

Is_Weekend 

In [163]:
df_fact_inventory.shape

(1500, 9)

In [164]:
df_fact_inventory.info()

<class 'pandas.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Snapshot_Date   1499 non-null   str    
 1   Warehouse_ID    1500 non-null   str    
 2   Product_ID      1500 non-null   str    
 3   On_Hand_Qty     1500 non-null   int64  
 4   Reserved_Qty    1499 non-null   float64
 5   In_Transit_Qty  1499 non-null   float64
 6   Reorder_Point   1499 non-null   float64
 7   Safety_Stock    1500 non-null   int64  
 8   Daily_Demand    1500 non-null   int64  
dtypes: float64(3), int64(3), str(3)
memory usage: 105.6 KB


In [165]:
df_fact_inventory.isnull().sum()

Snapshot_Date     1
Warehouse_ID      0
Product_ID        0
On_Hand_Qty       0
Reserved_Qty      1
In_Transit_Qty    1
Reorder_Point     1
Safety_Stock      0
Daily_Demand      0
dtype: int64

In [166]:
df_fact_inventory["Snapshot_Date"]=pd.to_datetime(df_fact_inventory["Snapshot_Date"],errors="coerce")

In [167]:
df_fact_inventory[df_fact_inventory.isna().any(axis=1)]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
399,NaT,LOC19,SKU028,118,76.0,89.0,103.0,35,12
700,2025-01-15,LOC21,SKU022,488,88.0,581.0,NaN,280,48
800,2025-12-03,LOC21,SKU044,217,NaN,522.0,233.0,122,36
1300,2025-11-19,LOC24,SKU032,187,125.0,NaN,84.0,31,15


In [168]:
df_fact_inventory[(df_fact_inventory["Warehouse_ID"]=="LOC19")&(df_fact_inventory["Product_ID"]=="SKU028")]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
105,2025-01-01,LOC19,SKU028,6,0.0,15.0,17.0,7,2
122,2025-11-19,LOC19,SKU028,30,7.0,48.0,48.0,25,5
132,2025-08-13,LOC19,SKU028,93,8.0,149.0,99.0,24,10
161,2025-03-12,LOC19,SKU028,39,25.0,57.0,21.0,21,4
203,2025-04-09,LOC19,SKU028,88,12.0,23.0,120.0,67,15
281,2025-10-22,LOC19,SKU028,65,8.0,43.0,45.0,37,7
283,2025-06-04,LOC19,SKU028,22,13.0,28.0,59.0,26,6
299,2025-07-02,LOC19,SKU028,100,7.0,77.0,72.0,28,9
399,NaT,LOC19,SKU028,118,76.0,89.0,103.0,35,12
424,2025-07-30,LOC19,SKU028,39,10.0,121.0,101.0,94,18


In [169]:
df_fact_inventory[
    (df_fact_inventory["Warehouse_ID"] == "LOC19") |
    (df_fact_inventory["Product_ID"] == "SKU028")
].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
1308,2025-01-01,LOC19,SKU030,167,50.0,292.0,233.0,163,32
271,2025-01-01,LOC19,SKU043,244,42.0,484.0,323.0,98,34
819,2025-01-01,LOC19,SKU012,175,113.0,4.0,169.0,57,22
1245,2025-01-01,LOC19,SKU019,115,61.0,95.0,146.0,61,27
105,2025-01-01,LOC19,SKU028,6,0.0,15.0,17.0,7,2
...,...,...,...,...,...,...,...,...,...
878,2025-12-03,LOC19,SKU012,79,38.0,68.0,86.0,29,11
302,2025-12-03,LOC19,SKU019,120,81.0,4.0,156.0,91,18
1269,2025-12-03,LOC19,SKU028,8,1.0,0.0,8.0,3,1
661,2025-12-03,LOC25,SKU028,193,104.0,22.0,170.0,81,20


In [170]:
df_fact_inventory.iloc[396:401][["Snapshot_Date"]]

,Snapshot_Date
396,2025-02-26
397,2025-11-05
398,2025-12-03
399,NaT
400,2025-08-13


In [171]:
df_fact_inventory[df_fact_inventory.isna().any(axis=1)]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
399,NaT,LOC19,SKU028,118,76.0,89.0,103.0,35,12
700,2025-01-15,LOC21,SKU022,488,88.0,581.0,NaN,280,48
800,2025-12-03,LOC21,SKU044,217,NaN,522.0,233.0,122,36
1300,2025-11-19,LOC24,SKU032,187,125.0,NaN,84.0,31,15


In [172]:
# missing reorder_point
df_fact_inventory[(df_fact_inventory["Warehouse_ID"]=="LOC21")&(df_fact_inventory["Product_ID"]=="SKU022")].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
700,2025-01-15,LOC21,SKU022,488,88.0,581.0,NaN,280,48
679,2025-01-29,LOC21,SKU022,227,26.0,63.0,254.0,73,29
1187,2025-02-12,LOC21,SKU022,113,49.0,294.0,212.0,49,23
82,2025-02-26,LOC21,SKU022,258,59.0,313.0,274.0,114,31
930,2025-03-12,LOC21,SKU022,243,136.0,39.0,181.0,103,32
901,2025-03-26,LOC21,SKU022,278,79.0,327.0,229.0,118,35
202,2025-04-09,LOC21,SKU022,380,200.0,271.0,259.0,269,46
733,2025-04-23,LOC21,SKU022,272,95.0,179.0,293.0,74,34
1346,2025-05-07,LOC21,SKU022,226,140.0,193.0,149.0,67,23
1447,2025-05-21,LOC21,SKU022,117,65.0,167.0,114.0,107,18


In [173]:
# missing In_Transit_Qty	
df_fact_inventory[(df_fact_inventory["Warehouse_ID"]=="LOC24")&(df_fact_inventory["Product_ID"]=="SKU032")].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
715,2025-01-01,LOC24,SKU032,54,3.0,74.0,36.0,23,6
745,2025-01-15,LOC24,SKU032,7,2.0,56.0,26.0,14,4
970,2025-01-29,LOC24,SKU032,147,27.0,89.0,165.0,71,17
973,2025-02-12,LOC24,SKU032,146,68.0,139.0,105.0,79,14
461,2025-02-26,LOC24,SKU032,31,5.0,52.0,71.0,66,12
215,2025-03-12,LOC24,SKU032,33,13.0,70.0,106.0,29,11
1277,2025-03-26,LOC24,SKU032,198,84.0,240.0,146.0,102,20
129,2025-04-09,LOC24,SKU032,233,125.0,113.0,124.0,90,20
77,2025-04-23,LOC24,SKU032,63,16.0,30.0,46.0,13,5
181,2025-05-07,LOC24,SKU032,97,42.0,148.0,88.0,47,11


In [174]:
# missing In_Reserved_Qty
df_fact_inventory[(df_fact_inventory["Warehouse_ID"]=="LOC21")&(df_fact_inventory["Product_ID"]=="SKU044")].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
135,2025-01-01,LOC21,SKU044,326,145.0,415.0,286.0,72,34
1005,2025-01-15,LOC21,SKU044,122,50.0,349.0,143.0,94,26
477,2025-01-29,LOC21,SKU044,489,100.0,441.0,207.0,134,40
497,2025-02-12,LOC21,SKU044,242,60.0,130.0,184.0,143,28
823,2025-02-26,LOC21,SKU044,174,12.0,302.0,207.0,96,31
1446,2025-03-12,LOC21,SKU044,350,71.0,167.0,335.0,124,34
548,2025-03-26,LOC21,SKU044,173,113.0,241.0,133.0,132,26
481,2025-04-09,LOC21,SKU044,112,67.0,130.0,213.0,104,27
939,2025-04-23,LOC21,SKU044,277,184.0,316.0,175.0,86,31
887,2025-05-07,LOC21,SKU044,31,18.0,26.0,185.0,72,20


In [175]:
df_fact_inventory.columns

Index(['Snapshot_Date', 'Warehouse_ID', 'Product_ID', 'On_Hand_Qty',
       'Reserved_Qty', 'In_Transit_Qty', 'Reorder_Point', 'Safety_Stock',
       'Daily_Demand'],
      dtype='str')

In [176]:
df_fact_inventory[
    (df_fact_inventory["On_Hand_Qty"]<0)|
    (df_fact_inventory["Reserved_Qty"]<0)|
    (df_fact_inventory["In_Transit_Qty"]<0)|
    (df_fact_inventory["Reorder_Point"]<0)|
    (df_fact_inventory["Safety_Stock"]<0)|
    (df_fact_inventory["Daily_Demand"]<0)
    ]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
30,2025-09-10,LOC22,SKU060,-25,37.0,138.0,112.0,95,16
185,2025-07-02,LOC24,SKU002,152,32.0,201.0,289.0,-20,33
600,2025-05-21,LOC20,SKU015,58,6.0,-40.0,42.0,36,8


In [177]:
df_fact_inventory[
    (df_fact_inventory["Warehouse_ID"] == "LOC22") &
    (df_fact_inventory["Product_ID"] == "SKU060")
].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
592,2025-01-01,LOC22,SKU060,197,90.0,305.0,257.0,138,26
85,2025-01-15,LOC22,SKU060,303,202.0,402.0,232.0,164,31
914,2025-01-29,LOC22,SKU060,33,9.0,25.0,50.0,25,6
976,2025-02-12,LOC22,SKU060,173,55.0,93.0,149.0,87,18
1165,2025-02-26,LOC22,SKU060,231,120.0,360.0,157.0,133,26
759,2025-03-12,LOC22,SKU060,216,86.0,69.0,132.0,113,23
46,2025-03-26,LOC22,SKU060,220,45.0,141.0,110.0,35,16
312,2025-04-09,LOC22,SKU060,202,7.0,302.0,278.0,162,34
450,2025-04-23,LOC22,SKU060,5000,23.0,113.0,200.0,111,23
720,2025-05-07,LOC22,SKU060,26,10.0,35.0,47.0,15,5


In [178]:
df_fact_inventory[
    (df_fact_inventory["Warehouse_ID"] == "LOC24") &
    (df_fact_inventory["Product_ID"] == "SKU002")
].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
769,2025-01-01,LOC24,SKU002,298,79.0,266.0,172.0,95,30
1088,2025-01-15,LOC24,SKU002,313,23.0,162.0,243.0,71,32
1093,2025-01-29,LOC24,SKU002,218,136.0,311.0,312.0,134,34
678,2025-02-12,LOC24,SKU002,197,113.0,407.0,348.0,178,35
648,2025-02-26,LOC24,SKU002,243,103.0,333.0,224.0,164,31
549,2025-03-12,LOC24,SKU002,418,211.0,63.0,210.0,221,38
629,2025-03-26,LOC24,SKU002,199,17.0,315.0,158.0,123,26
225,2025-04-09,LOC24,SKU002,97,13.0,218.0,302.0,206,35
479,2025-04-23,LOC24,SKU002,124,22.0,18.0,111.0,37,17
585,2025-05-07,LOC24,SKU002,334,175.0,218.0,220.0,193,40


In [179]:
df_fact_inventory[
    (df_fact_inventory["Warehouse_ID"] == "LOC20") &
    (df_fact_inventory["Product_ID"] == "SKU015")
].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
652,2025-01-01,LOC20,SKU015,222,70.0,175.0,159.0,91,28
804,2025-01-15,LOC20,SKU015,123,79.0,79.0,129.0,83,23
1295,2025-01-29,LOC20,SKU015,161,29.0,242.0,237.0,72,25
712,2025-02-12,LOC20,SKU015,147,67.0,41.0,177.0,127,23
598,2025-02-26,LOC20,SKU015,39,21.0,230.0,221.0,54,26
751,2025-03-12,LOC20,SKU015,159,25.0,44.0,116.0,106,21
650,2025-03-26,LOC20,SKU015,234,1.0,135.0,203.0,76,22
2,2025-04-09,LOC20,SKU015,192,6.0,88.0,184.0,87,26
178,2025-04-23,LOC20,SKU015,245,140.0,99.0,196.0,122,22
597,2025-05-07,LOC20,SKU015,115,39.0,272.0,131.0,46,20


#### Current fact_inventory findings

Missing values: 4 → unresolved because they cannot be reliably reconstructed.

Negative values: 3 → identified and unresolved because their correct values cannot be established.

In [180]:
df_fact_inventory[df_fact_inventory.duplicated(subset=["Snapshot_Date","Warehouse_ID","Product_ID"],keep=False)]


,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand


In [181]:
df_fact_inventory[
    ~ df_fact_inventory["Warehouse_ID"].isin(
        df_dim_location["Location_ID"]
    )
]["Warehouse_ID"].value_counts()

Warehouse_ID
LOC99    1
loc19    1
Name: count, dtype: int64

In [182]:
df_fact_inventory.loc[df_fact_inventory["Warehouse_ID"]=="loc19","Warehouse_ID"]="LOC19"

In [183]:
df_fact_inventory[
    ~df_fact_inventory["Warehouse_ID"].isin(
        df_dim_location["Location_ID"]
    )
]["Warehouse_ID"].value_counts()

Warehouse_ID
LOC99    1
Name: count, dtype: int64

In [184]:
df_fact_inventory["Warehouse_ID"].value_counts().loc[["LOC19"]]

Warehouse_ID
LOC19    150
Name: count, dtype: int64

In [185]:
df_fact_inventory[df_fact_inventory["Warehouse_ID"]=="LOC99"]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
310,2025-04-09,LOC99,SKU012,105,64.0,70.0,84.0,30,13


#### Unresolved foreign-key anomaly: LOC99 appears once in fact_inventory but does not exist in dim_location. No reliable evidence was found to map it to a valid location.

In [186]:
df_fact_inventory[
    ~ df_fact_inventory["Product_ID"].isin(
        df_dim_product["Product_ID"]
    )
]["Product_ID"].value_counts()

Product_ID
SKU999      1
 sku010     1
Name: count, dtype: int64

In [187]:
df_dim_product[df_dim_product["Product_ID"].str.lower()=="sku010"]

,Product_ID,SKU,Product_Name,Category,Unit_Cost_USD,Unit_Selling_Price_USD,Criticality
9,SKU010,CMP-010,Memory Module,Components,969.85,1300.54,Medium


In [188]:
df_fact_inventory[df_fact_inventory["Product_ID"]=="sku010"]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand


In [189]:
df_fact_inventory[
    ~ df_fact_inventory["Product_ID"].isin(
        df_dim_product["Product_ID"]
    )
]["Product_ID"].map(repr)

240      'SKU999'
900    ' sku010 '
Name: Product_ID, dtype: str

In [190]:
df_fact_inventory["Product_ID"]=df_fact_inventory["Product_ID"].str.strip().str.upper()


In [191]:
df_fact_inventory[
    ~df_fact_inventory["Product_ID"].isin(
        df_dim_product["Product_ID"]
    )
]["Product_ID"].value_counts()

Product_ID
SKU999    1
Name: count, dtype: int64

In [192]:
df_fact_inventory[df_fact_inventory["Product_ID"]=="SKU999"]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
240,2025-05-21,LOC22,SKU999,176,93.0,117.0,246.0,221,37


In [193]:
df_fact_inventory.iloc[238:242]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
238,2025-10-08,LOC24,SKU012,418,102.0,84.0,217.0,109,32
239,2025-12-03,LOC21,SKU056,129,43.0,88.0,92.0,41,12
240,2025-05-21,LOC22,SKU999,176,93.0,117.0,246.0,221,37
241,2025-08-27,LOC25,SKU024,15,4.0,51.0,110.0,51,12


#### Unresolved product-key anomaly: SKU999 appears once in fact_inventory but does not exist in dim_product. No reliable evidence was found to map it to a valid product, so the value was not replaced.

In [194]:
df_fact_inventory.columns

Index(['Snapshot_Date', 'Warehouse_ID', 'Product_ID', 'On_Hand_Qty',
       'Reserved_Qty', 'In_Transit_Qty', 'Reorder_Point', 'Safety_Stock',
       'Daily_Demand'],
      dtype='str')

In [195]:
df_fact_inventory[(df_fact_inventory["On_Hand_Qty"]<0)|
                  (df_fact_inventory["Reserved_Qty"]<0)|
                  (df_fact_inventory["In_Transit_Qty"]<0)|
                  (df_fact_inventory["Reorder_Point"]<0)|
                  (df_fact_inventory["Safety_Stock"]<0)|
                  (df_fact_inventory["Daily_Demand"]<0)]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
30,2025-09-10,LOC22,SKU060,-25,37.0,138.0,112.0,95,16
185,2025-07-02,LOC24,SKU002,152,32.0,201.0,289.0,-20,33
600,2025-05-21,LOC20,SKU015,58,6.0,-40.0,42.0,36,8


In [196]:
df_fact_inventory[(df_fact_inventory["Warehouse_ID"]=="LOC22")&(df_fact_inventory["Product_ID"]=="SKU060")]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
30,2025-09-10,LOC22,SKU060,-25,37.0,138.0,112.0,95,16
46,2025-03-26,LOC22,SKU060,220,45.0,141.0,110.0,35,16
85,2025-01-15,LOC22,SKU060,303,202.0,402.0,232.0,164,31
192,2025-11-05,LOC22,SKU060,83,32.0,115.0,162.0,73,18
201,2025-05-21,LOC22,SKU060,196,55.0,346.0,203.0,116,24
229,2025-11-19,LOC22,SKU060,72,7.0,60.0,168.0,60,18
264,2025-12-03,LOC22,SKU060,209,64.0,123.0,177.0,50,22
297,2025-10-22,LOC22,SKU060,223,37.0,120.0,129.0,59,23
312,2025-04-09,LOC22,SKU060,202,7.0,302.0,278.0,162,34
391,2025-08-13,LOC22,SKU060,97,25.0,6.0,160.0,41,19


#### Historical snapshots were reviewed, but the correct values could not be reliably reconstructed. Therefore, the anomalous values were retained and flagged rather than replaced with assumed values.

In [197]:
df_fact_inventory[(df_fact_inventory["Reserved_Qty"])>df_fact_inventory["On_Hand_Qty"]]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
30,2025-09-10,LOC22,SKU060,-25,37.0,138.0,112.0,95,16
71,2025-07-16,LOC23,SKU018,95,145.0,185.0,117.0,45,14
99,2025-03-26,LOC24,SKU025,0,1.0,271.0,273.0,227,43
273,2025-12-03,LOC25,SKU046,0,1.0,92.0,142.0,107,19
290,2025-05-07,LOC21,SKU030,0,1.0,39.0,108.0,45,14
482,2025-07-16,LOC24,SKU002,0,1.0,373.0,175.0,126,28
491,2025-07-30,LOC25,SKU058,0,1.0,59.0,63.0,24,7
908,2025-05-21,LOC25,SKU058,0,1.0,167.0,182.0,67,25
920,2025-12-03,LOC22,SKU004,0,1.0,115.0,184.0,113,19
1009,2025-03-26,LOC24,SKU012,0,1.0,102.0,50.0,26,9


In [198]:
df_fact_inventory[(df_fact_inventory["Warehouse_ID"] == "LOC23") &(df_fact_inventory["Product_ID"] == "SKU018")].sort_values("Snapshot_Date")

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand
559,2025-01-01,LOC23,SKU018,78,30.0,32.0,181.0,115,21
682,2025-01-15,LOC23,SKU018,238,43.0,139.0,146.0,80,26
971,2025-01-29,LOC23,SKU018,276,151.0,379.0,209.0,148,27
206,2025-02-12,LOC23,SKU018,104,57.0,76.0,50.0,24,8
1469,2025-02-26,LOC23,SKU018,188,67.0,101.0,134.0,60,18
411,2025-03-12,LOC23,SKU018,48,21.0,88.0,55.0,20,6
862,2025-03-26,LOC23,SKU018,124,84.0,74.0,188.0,83,19
25,2025-04-09,LOC23,SKU018,276,111.0,170.0,228.0,99,24
1402,2025-04-23,LOC23,SKU018,129,87.0,22.0,128.0,75,21
468,2025-05-07,LOC23,SKU018,63,13.0,74.0,70.0,57,14


In [199]:
df_fact_inventory.dtypes


Snapshot_Date     datetime64[us]
Warehouse_ID                 str
Product_ID                   str
On_Hand_Qty                int64
Reserved_Qty             float64
In_Transit_Qty           float64
Reorder_Point            float64
Safety_Stock               int64
Daily_Demand               int64
dtype: object

In [200]:
df_fact_inventory.isna().sum()

Snapshot_Date     1
Warehouse_ID      0
Product_ID        0
On_Hand_Qty       0
Reserved_Qty      1
In_Transit_Qty    1
Reorder_Point     1
Safety_Stock      0
Daily_Demand      0
dtype: int64

In [201]:
df_fact_inventory.duplicated().sum()

np.int64(0)

In [202]:
# df_fact_shipments - df_dim_supplier

df_fact_shipments[
    ~ df_fact_shipments["Supplier_ID"].isin(
        df_dim_supplier["Supplier_ID"]
    )
]["Supplier_ID"].value_counts()

Series([], Name: count, dtype: int64)

In [203]:
#df_fact_shipments - df_dim_product

df_fact_shipments[
    ~ df_fact_shipments["Product_ID"].isin(
        df_dim_product["Product_ID"]
    )
]["Product_ID"].value_counts()

Series([], Name: count, dtype: int64)

In [204]:
# df_fact_shipments["Origin_ID"] - df_dim_location

df_fact_shipments[
    ~ df_fact_shipments["Origin_ID"].isin(
        df_dim_location["Location_ID"]
    )
]["Origin_ID"].value_counts()

Series([], Name: count, dtype: int64)

In [205]:
# df_fact_shipments["Destination_ID"] - df_dim_location

df_fact_shipments[
    ~ df_fact_shipments["Destination_ID"].isin(
        df_dim_location["Location_ID"]
    )
]["Destination_ID"].value_counts()

Series([], Name: count, dtype: int64)

In [206]:
# df_fact_shipments["Ship_Date"] - df_dim_date

df_fact_shipments[
    ~ df_fact_shipments["Ship_Date"].isin(
        df_dim_date["Date"]
    )
]["Ship_Date"].value_counts()

Series([], Name: count, dtype: int64)

In [207]:
# df_fact_shipments["Expected_Date"] - df_dim_date

df_fact_shipments[
    ~ df_fact_shipments["Expected_Date"].isin(
        df_dim_date["Date"]
    )
]["Expected_Date"].value_counts()

Series([], Name: count, dtype: int64)

In [208]:
# df_fact_shipments["Actual_Delivery_Date"] - df_dim_date

df_fact_shipments[
    ~ df_fact_shipments["Actual_Delivery_Date"].isin(
        df_dim_date["Date"]
    )
]["Actual_Delivery_Date"].value_counts()

Series([], Name: count, dtype: int64)

In [209]:
# df_fact_inventory["Snapshot_Date"] - df_dim_date

df_fact_inventory[
    ~ df_fact_inventory["Snapshot_Date"].isin(
        df_dim_date["Date"]
    )
]["Snapshot_Date"].value_counts()

Series([], Name: count, dtype: int64)

In [210]:
# df_fact_inventory["Warehouse_ID"] - df_dim_location

df_fact_inventory[
    ~ df_fact_inventory["Warehouse_ID"].isin(
        df_dim_location["Location_ID"]
    )
]["Warehouse_ID"].value_counts()

Warehouse_ID
LOC99    1
Name: count, dtype: int64

In [211]:
# df_fact_inventory["Product_ID"] - df_dim_product

df_fact_inventory[
    ~ df_fact_inventory["Product_ID"].isin(
        df_dim_product["Product_ID"]
    )
]["Product_ID"].value_counts()

Product_ID
SKU999    1
Name: count, dtype: int64

In [212]:
df_dim_supplier["Supplier_ID"].duplicated().sum()

np.int64(0)

In [213]:
df_dim_product["Product_ID"].duplicated().sum()

np.int64(0)

In [214]:
df_dim_location["Location_ID"].duplicated().sum()

np.int64(0)

In [215]:
df_dim_date["Date"].duplicated().sum()

np.int64(0)

In [216]:
print("Kernel is working")

Kernel is working


In [217]:
import sys

print(sys.version)
print(sys.executable)

3.14.4 (tags/v3.14.4:23116f9, Apr  7 2026, 14:10:54) [MSC v.1944 64 bit (AMD64)]
c:\Users\MY\AppData\Local\Programs\Python\Python314\python.exe


In [218]:
import sqlalchemy

print(sqlalchemy.__version__)

2.0.52


In [219]:
import mysql.connector

print("MySQL connector loaded successfully")

MySQL connector loaded successfully


In [220]:
import os

os.makedirs("sql_export", exist_ok=True)

df_dim_date.to_csv("sql_export/dim_date.csv", index=False)
df_dim_product.to_csv("sql_export/dim_product.csv", index=False)
df_dim_location.to_csv("sql_export/dim_location.csv", index=False)
df_dim_supplier.to_csv("sql_export/dim_supplier.csv", index=False)
df_fact_shipments.to_csv("sql_export/fact_shipments.csv", index=False)
df_fact_inventory.to_csv("sql_export/fact_inventory.csv", index=False)

print("All tables exported successfully.")

All tables exported successfully.


In [221]:
df_dim_date_sql = df_dim_date.copy()

df_dim_date_sql["Is_Weekend"] = df_dim_date_sql["Is_Weekend"].astype(int)

df_dim_date_sql.to_csv(
    "sql_export/dim_date.csv",
    index=False
)

print(df_dim_date_sql["Is_Weekend"].value_counts())

Is_Weekend
0    522
1    208
Name: count, dtype: int64


In [222]:
print(df_dim_date_sql.head())
print(df_dim_date_sql["Is_Weekend"].unique())

        Date  Year  Month Month_Name Quarter  Week  Day   Day_Name  Is_Weekend
0 2025-01-01  2025      1    January      Q1     1    1  Wednesday           0
1 2025-01-02  2025      1    January      Q1     1    2   Thursday           0
2 2025-01-03  2025      1    January      Q1     1    3     Friday           0
3 2025-01-04  2025      1    January      Q1     1    4   Saturday           1
4 2025-01-05  2025      1    January      Q1     1    5     Sunday           1
[0 1]


In [223]:
import os

print(os.path.abspath("sql_export/dim_date.csv"))

c:\Users\MY\Downloads\sql_export\dim_date.csv


In [224]:
import pandas as pd

test = pd.read_csv("sql_export/dim_date.csv")

print("Rows:", len(test))
print(test.head())

Rows: 730
         Date  Year  Month Month_Name Quarter  Week  Day   Day_Name  \
0  2025-01-01  2025      1    January      Q1     1    1  Wednesday   
1  2025-01-02  2025      1    January      Q1     1    2   Thursday   
2  2025-01-03  2025      1    January      Q1     1    3     Friday   
3  2025-01-04  2025      1    January      Q1     1    4   Saturday   
4  2025-01-05  2025      1    January      Q1     1    5     Sunday   

   Is_Weekend  
0           0  
1           0  
2           0  
3           1  
4           1  


In [225]:
print(os.path.abspath("sql_export/dim_date.csv"))
print("Rows:", len(test))

c:\Users\MY\Downloads\sql_export\dim_date.csv
Rows: 730


#### Paused the MySQL implementation due to compatibility and import issues between Python 3.14, MySQL Connector, and MySQL Workbench

#### Pre-Power BI checklist

##### Data quality Checks — already mostly done (will be adding the aswers to these in readme)

2.What missing values exist?

3.Are there duplicate records?

4.Are there invalid/inconsistent Product or Warehouse IDs?

5.Are there impossible values?

6.Which anomalies cannot be reconstructed and must be documented?

Already covered most of this.

In [226]:
df_fact_inventory.columns

Index(['Snapshot_Date', 'Warehouse_ID', 'Product_ID', 'On_Hand_Qty',
       'Reserved_Qty', 'In_Transit_Qty', 'Reorder_Point', 'Safety_Stock',
       'Daily_Demand'],
      dtype='str')

In [228]:
# 6.How much inventory is available?

df_fact_inventory["Available_Qty"]=df_fact_inventory["On_Hand_Qty"]-df_fact_inventory["Reserved_Qty"]
df_fact_inventory["Available_Qty"].describe()

count    1499.000000
mean       96.719813
std       146.698974
min       -62.000000
25%        40.000000
50%        78.000000
75%       131.000000
max      4977.000000
Name: Available_Qty, dtype: float64

##### Some records have negative Available_Qty, meaning Reserved_Qty exceeds
On_Hand_Qty. This may represent an operational inventory situation and
requires further investigation.

In [230]:
# 7. Which records are at stockout risk?

df_fact_inventory[df_fact_inventory["On_Hand_Qty"]<=0]

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand,Available_Qty
30,2025-09-10,LOC22,SKU060,-25,37.0,138.0,112.0,95,16,-62.0
83,2025-07-02,LOC19,SKU040,0,0.0,80.0,89.0,19,9,0.0
99,2025-03-26,LOC24,SKU025,0,1.0,271.0,273.0,227,43,-1.0
273,2025-12-03,LOC25,SKU046,0,1.0,92.0,142.0,107,19,-1.0
290,2025-05-07,LOC21,SKU030,0,1.0,39.0,108.0,45,14,-1.0
482,2025-07-16,LOC24,SKU002,0,1.0,373.0,175.0,126,28,-1.0
491,2025-07-30,LOC25,SKU058,0,1.0,59.0,63.0,24,7,-1.0
792,2025-07-30,LOC22,SKU060,0,0.0,256.0,187.0,47,20,0.0
834,2025-01-29,LOC23,SKU026,0,0.0,192.0,170.0,102,24,0.0
853,2025-03-12,LOC21,SKU030,0,0.0,22.0,105.0,47,15,0.0


##### 21 inventory records had On_Hand_Qty <= 0. These include zero-stock situations as well as one negative on-hand observation. Several zero-stock records had inventory in transit, suggesting that replenishment may already have been underway. Some records also had reservations despite zero on-hand inventory, resulting in negative available inventory. These situations should be treated as inventory-risk events rather than automatically classified as data errors.

In [231]:
# 8. Products below their reorder point

df_reorder_risk = df_fact_inventory[df_fact_inventory["On_Hand_Qty"] <= df_fact_inventory["Reorder_Point"]]

df_reorder_risk.shape

(851, 10)

In [233]:
df_reorder_risk[["Snapshot_Date","Warehouse_ID","Product_ID","On_Hand_Qty","In_Transit_Qty","Reorder_Point","Daily_Demand"]].head(10)

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,In_Transit_Qty,Reorder_Point,Daily_Demand
0,2025-09-10,LOC23,SKU017,70,103.0,76.0,8
1,2025-11-05,LOC23,SKU042,226,398.0,263.0,28
5,2025-07-16,LOC25,SKU051,73,59.0,119.0,14
8,2025-10-22,LOC23,SKU018,92,34.0,144.0,16
9,2025-02-26,LOC24,SKU040,27,40.0,32.0,4
10,2025-07-16,LOC23,SKU042,77,53.0,91.0,14
17,2025-05-21,LOC25,SKU046,93,79.0,160.0,18
19,2025-01-01,LOC20,SKU059,81,313.0,147.0,23
21,2025-09-10,LOC23,SKU042,100,135.0,174.0,24
22,2025-09-24,LOC24,SKU060,23,12.0,53.0,8


##### 851 of 1,499 inventory records (56.8%) were at or below their reorder point, indicating that a substantial proportion of inventory observations were in a replenishment-risk state. However, several of these records had inventory already in transit, so reorder-point risk should be evaluated alongside incoming inventory and demand before concluding that a shortage exists.

In [234]:
# 9. Safety-stock risk

df_safety_risk = df_fact_inventory[df_fact_inventory["Available_Qty"] <= df_fact_inventory["Safety_Stock"]]
df_safety_risk.shape

(657, 10)

In [235]:
df_safety_risk.head(10)

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand,Available_Qty
1,2025-11-05,LOC23,SKU042,226,95.0,398.0,263.0,153,28,131.0
3,2025-03-26,LOC19,SKU019,124,49.0,194.0,108.0,113,19,75.0
5,2025-07-16,LOC25,SKU051,73,16.0,59.0,119.0,60,14,57.0
6,2025-08-27,LOC22,SKU059,40,23.0,31.0,29.0,20,5,17.0
8,2025-10-22,LOC23,SKU018,92,50.0,34.0,144.0,50,16,42.0
21,2025-09-10,LOC23,SKU042,100,39.0,135.0,174.0,142,24,61.0
22,2025-09-24,LOC24,SKU060,23,10.0,12.0,53.0,17,8,13.0
23,2025-10-08,LOC23,SKU028,87,42.0,178.0,79.0,65,12,45.0
24,2025-11-19,LOC23,SKU034,40,19.0,230.0,132.0,65,24,21.0
27,2025-12-03,LOC20,SKU057,136,90.0,180.0,143.0,52,18,46.0


##### 657 of 1,499 inventory records (43.8%) had available inventory at or below their safety-stock level, indicating that a significant proportion of observations were operating with limited inventory buffers. Several of these records had inventory in transit, suggesting that replenishment may already have been underway. Therefore, safety-stock risk should be interpreted together with incoming inventory and demand rather than treated as an immediate stockout.

In [246]:
# 10. Days of inventory coverage

import numpy as np

df_fact_inventory["Inventory_Risk"]=np.select([
    df_fact_inventory["Available_Qty"]<0,
    df_fact_inventory["Daily_Demand"]==0,
    df_fact_inventory["Days_of_Cover"]==2,
    df_fact_inventory["Days_of_Cover"]==5,
    df_fact_inventory["Days_of_Cover"]==10
    ],
    [
        "Critical",
        "Not Applicable",
        "Critical",
        "Low",
        "Medium"
    ],
    default="Healthy"
    )
df_fact_inventory["Inventory_Risk"].value_counts(dropna=False)

Inventory_Risk
Healthy           1433
Critical            36
Low                 28
Medium               2
Not Applicable       1
Name: count, dtype: int64

In [258]:
df_fact_inventory.head(20)

,Snapshot_Date,Warehouse_ID,Product_ID,On_Hand_Qty,Reserved_Qty,In_Transit_Qty,Reorder_Point,Safety_Stock,Daily_Demand,Available_Qty,Days_of_Cover,Inventory_Risk
0,2025-09-10,LOC23,SKU017,70,3.0,103.0,76.0,42,8,67.0,8.375000,Healthy
1,2025-11-05,LOC23,SKU042,226,95.0,398.0,263.0,153,28,131.0,4.678571,Healthy
2,2025-04-09,LOC20,SKU015,192,6.0,88.0,184.0,87,26,186.0,7.153846,Healthy
3,2025-03-26,LOC19,SKU019,124,49.0,194.0,108.0,113,19,75.0,3.947368,Healthy
4,2025-04-09,LOC20,SKU024,136,31.0,264.0,134.0,74,22,105.0,4.772727,Healthy
5,2025-07-16,LOC25,SKU051,73,16.0,59.0,119.0,60,14,57.0,4.071429,Healthy
6,2025-08-27,LOC22,SKU059,40,23.0,31.0,29.0,20,5,17.0,3.400000,Healthy
7,2025-06-18,LOC25,SKU051,77,6.0,58.0,60.0,27,11,71.0,6.454545,Healthy
8,2025-10-22,LOC23,SKU018,92,50.0,34.0,144.0,50,16,42.0,2.625000,Healthy
9,2025-02-26,LOC24,SKU040,27,3.0,40.0,32.0,13,4,24.0,6.000000,Healthy


##### The inventory-risk classification identified 36 Critical, 28 Low, 2 Medium, 1,433 Healthy, and 1 Not Applicable observations. The majority of observations were classified as Healthy under the current Days-of-Cover thresholds. However, reorder-point and safety-stock checks identified substantially more risk observations, indicating that inventory risk depends on multiple operational indicators rather than Days of Cover alone.

#### Product & Warehouse summary
1. Risk by warehouse

In [262]:
df_fact_inventory.groupby("Warehouse_ID")["Inventory_Risk"].value_counts()

Warehouse_ID  Inventory_Risk
LOC19         Healthy           145
              Critical            4
              Low                 1
LOC20         Healthy           238
              Low                 9
              Critical            2
              Medium              1
LOC21         Healthy           165
              Critical            5
              Low                 5
LOC22         Healthy           237
              Critical            6
              Low                 5
              Medium              1
LOC23         Healthy           194
              Critical            4
              Low                 2
LOC24         Healthy           239
              Critical            8
              Low                 3
LOC25         Healthy           214
              Critical            7
              Low                 3
              Not Applicable      1
LOC99         Healthy             1
Name: count, dtype: int64

In [263]:
df_fact_inventory.groupby("Product_ID")["Inventory_Risk"].value_counts()

Product_ID  Inventory_Risk
SKU002      Healthy           47
            Critical           2
            Low                1
SKU004      Healthy           48
            Low                1
                              ..
SKU059      Low                5
            Critical           3
SKU060      Healthy           49
            Critical           1
SKU999      Healthy            1
Name: count, Length: 74, dtype: int64

##### Inventory risk was observed across the main warehouse locations, with both Critical and Low risk records appearing at multiple warehouses. LOC24 and LOC25 had relatively more Critical observations in the displayed results, while LOC20 had more Low-risk observations. LOC99 contained only one Healthy observation and should be interpreted cautiously because of its very small number of records.
##### Inventory risk was distributed across multiple products. SKU059 showed several risk observations, including 3 Critical and 5 Low records, while SKU002 had 2 Critical and 1 Low record. SKU999 appeared as a single Healthy observation but remains an unresolved product-key anomaly identified during data-quality validation.

In [264]:
df_fact_shipments.columns

Index(['Shipment_ID', 'Order_ID', 'Supplier_ID', 'Product_ID', 'Origin_ID',
       'Destination_ID', 'Ship_Date', 'Expected_Date', 'Actual_Delivery_Date',
       'Shipping_Mode', 'Quantity', 'Freight_Cost_USD', 'Port_Dwell_Days',
       'Customs_Delay_Days', 'Transit_Time_Days', 'Supplier_Risk_Score',
       'Priority_Level', 'Shipment_Status', 'Revenue_Exposure_USD',
       'Delivery_Performance'],
      dtype='str')

#### Shipment analysis
1.How many shipments are there and what are their statuses?

In [269]:
df_fact_shipments["Shipment_Status"].value_counts()

Shipment_Status
Delivered    2700
delivered       1
Name: count, dtype: int64

In [270]:
df_fact_shipments["Shipment_Status"]=df_fact_shipments["Shipment_Status"].str.strip().str.title()


In [271]:
df_fact_shipments["Shipment_Status"].value_counts()

Shipment_Status
Delivered    2701
Name: count, dtype: int64

##### All 2,701 shipment records in the dataset are marked as Delivered. No alternative shipment statuses such as In Transit, Delayed, or Cancelled are present. Therefore, shipment status cannot be used to compare operational outcomes in this dataset, and delivery performance should instead be evaluated using the available delivery-performance and timing fields.

In [272]:
# 2. Delivery performance
df_fact_shipments["Delivery_Performance"].value_counts()

Delivery_Performance
late       2460
on_time     240
unknown       1
Name: count, dtype: int64

##### Of the 2,701 shipment records, 2,460 (91.1%) were classified as late, while only 240 (8.9%) were classified as on time. One shipment had an unknown delivery-performance value. This indicates a strong concentration of late deliveries in the dataset and warrants further investigation into transit time, port dwell, customs delays, shipping mode, and supplier-related factors.

In [273]:
# 3. Average transit time
df_fact_shipments["Transit_Time_Days"].describe()

count    2701.000000
mean       24.605702
std         8.850383
min         3.000000
25%        18.000000
50%        25.000000
75%        32.000000
max        48.000000
Name: Transit_Time_Days, dtype: float64

##### The average transit time was 24.61 days, with a median of 25 days across 2,701 shipments. Transit times ranged from 3 to 48 days, indicating substantial variation in shipment duration. The middle 50% of shipments had transit times between 18 and 32 days.

In [275]:
# 4. Shipping modes
df_fact_shipments["Shipping_Mode"].value_counts()

Shipping_Mode
Ocean    1601
Air       673
Rail      427
Name: count, dtype: int64

##### Ocean was the most frequently used shipping mode, accounting for 1,601 of 2,701 shipments (59.3%), followed by Air at 673 (24.9%) and Rail at 427 (15.8%). Ocean therefore represents the largest share of shipment activity in the dataset.

In [276]:
# 5. Port and customs delays
df_fact_shipments[["Port_Dwell_Days","Customs_Delay_Days"]].describe()

,Port_Dwell_Days,Customs_Delay_Days
count,2701.000000,2700.000000
mean,4.428360,2.483333
std,2.512147,2.215698
min,0.000000,0.000000
25%,3.000000,1.000000
50%,4.000000,2.000000
75%,6.000000,4.000000
max,14.000000,65.000000


##### Average port dwell time was 4.43 days, while average customs delay was 2.48 days. Port dwell ranged from 0 to 14 days. Customs delays ranged from 0 to 65 days, with one missing value among the 2,701 shipment records. The maximum customs delay is substantially higher than the typical values and should be investigated as a potential exceptional event rather than treated as representative of normal customs performance.

In [277]:
df_dim_date.to_csv(
    r"C:\Users\MY\Downloads\powerbi_data\dim_date.csv",
    index=False
)

In [278]:
import os

os.path.exists(
    r"C:\Users\MY\Downloads\powerbi_data\dim_date.csv"
)

True

In [279]:
df_dim_product.to_csv(
    r"C:\Users\MY\Downloads\powerbi_data\dim_product.csv",
    index=False
)

In [280]:
os.path.exists(
    r"C:\Users\MY\Downloads\powerbi_data\dim_product.csv"
)

True

In [281]:
df_dim_location.to_csv(
    r"C:\Users\MY\Downloads\powerbi_data\dim_location.csv",
    index=False
)

In [282]:
os.path.exists(
    r"C:\Users\MY\Downloads\powerbi_data\dim_location.csv"
)

True

In [283]:
df_fact_inventory.to_csv(
    r"C:\Users\MY\Downloads\powerbi_data\fact_inventory.csv",
    index=False
)

In [284]:
os.path.exists(
    r"C:\Users\MY\Downloads\powerbi_data\fact_inventory.csv"
)

True

In [285]:
df_fact_shipments.to_csv(
    r"C:\Users\MY\Downloads\powerbi_data\fact_shipments.csv",
    index=False
)

In [286]:
os.path.exists(
    r"C:\Users\MY\Downloads\powerbi_data\fact_shipments.csv"
)

True

In [287]:
import os

folder = r"C:\Users\MY\Downloads\powerbi_data"

os.listdir(folder)

['dim_date.csv',
 'dim_location.csv',
 'dim_product.csv',
 'fact_inventory.csv',
 'fact_shipments.csv']

In [288]:
print("dim_date:", len(df_dim_date))
print("dim_product:", len(df_dim_product))
print("dim_location:", len(df_dim_location))
print("fact_inventory:", len(df_fact_inventory))
print("fact_shipments:", len(df_fact_shipments))

dim_date: 730
dim_product: 60
dim_location: 25
fact_inventory: 1500
fact_shipments: 2701
